lib/src/source_visitor.dart

```diff
@@ -621,8 +609,90 @@ class SourceVisitor extends ThrowingAstVisitor {
 
     builder.endBlockArgumentNesting();
     builder.unnest();
+  }
+
+  /// Format the cascade using a nested block instead of a single inline
+  /// expression.
+  ///
+  /// If the cascade has multiple sections, we know each section will be on its
+  /// own line and we know there will be at least one trailing section following
+  /// a preceding one. That let's us treat all of the earlier sections as a
+  /// separate block like we do with collections and functions, instead of a
+  /// monolithic expression. Using a block in turn makes big cascades much
+  /// faster to format (like 10x) since the block formatting is memoized and
+  /// each cascade section in it is formatted independently.
+  ///
+  /// The tricky part is that block formatting assumes the entire line will be
+  /// part of the block. This is not true of the last section in a cascade,
+  /// which may have other trailing code, like the `;` here:
+  ///
+  ///     var x = someLeadingExpression
+  ///       ..firstCascade()
+  ///       ..secondCascade()
+  ///       ..thirdCascade()
+  ///       ..fourthCascade();
+  ///
+  /// To handle that, we don't put the last section in the block and instead
+  /// format it with the surrounding expression. So, from the formatter's
+  /// view, a cascade like:
+  ///
+  ///     var x = someLeadingExpression
+  ///       ..firstCascade()
+  ///       ..secondCascade()
+  ///       ..thirdCascade()
+  ///       ..fourthCascade();
+  ///
+  /// Is formatted like:
+  ///
+  ///     var x = someLeadingExpression
+  ///       [ begin block ]
+  ///       ..firstCascade()
+  ///       ..secondCascade()
+  ///       ..thirdCascade()
+  ///       [ end block ]
+  ///       ..fourthCascade();
+  ///
+  /// This somewhere between clever and hacky, but it works and allows cascades
+  /// of essentially unbounded length to be formatted quickly.
+  void _visitSplitCascade(CascadeExpression node) {
+    // Rule to split before the first `..`.
+    builder.startLazyRule(Rule.hard());
+    visit(node.target);
+
+    builder.nestExpression(indent: Indent.cascade, now: true);
+    builder.startBlockArgumentNesting();
+
+    zeroSplit();
+
+    // If there are comments before the first section, keep them outside of the
```

Comment:
Why is this necessary?

Reply:
The way nested blocks work is a little inelegant. It's part of the formatter that evolved over time and is more bolted on than I would like. Say you have code like:

```dart
var list = [ // hi
  one,
  two,
  three
];
```

Then the `// hi` gets claimed by the inner block and formatted like:

```dart
var list = [
  [ begin block ]
  // hi
  one,
  two,
  three
  [ end block ]
];
```

That generally looks right for collection literals and lambdas (which is what the nested chunk stuff was added for), but no for cascades. If you have:

```dart
receiver // hi
  ..cascade()
  ..another();
```

We want it to start formatted like that and not end up as:

```dart
receiver
  // hi
  ..cascade()
  ..another();
```

Keeping the comment outside of the block preserves that. Added a comment to explain.

(The formatter's codebase is not super heavy on tech debt, but there is definitely a simpler formatter hiding inside it trying to get out that I would love to surface if I had the ability to simplify some of the formatting rules and the time to do a massive refactor.)